### ISSUE TITLE
ArgumentError: invalid base64

### ISSUE THREAD

**Issue by subhashsaran** [zaru/webpush#22]
When i am trigerring push notifications, in case of chrome they are delievering successfully but in case of firefox they are always giving invalid base64 erro,. after this
```
 Webpush.payload_send(
        endpoint: device.endpoint,
        p256dh: device.p256dh,
        auth: device.auth,
        cert_store: OpenSSL::X509::Store.new.tap(&:set_default_paths),
        message: message_json(notification),

        api_key: ENV["GCM_KEY"]
      )

```
Please suggest some solution.

**Comment by rossta** [comment]
@subhashsaran Could you provide more info about the message you're receiving, like copy & pasting a stack trace (or is that a message you're getting from Firefox)? Have you tried refreshing the `p256dh` and `auth` keys? Does the same thing happen if you pass a simple message (we can't tell what `message_json(notification)` does)?

**Comment by subhashsaran** [comment]
message_json simply returns json encoded message object. same message is getting delivered to chrome subscibers but in case of firefox subscriber it's giving invalid base64 error.
```
def message_json(notification)
    JSON.generate(
      title: title,
      body: message,
      icon: notification.image_attachment.url,
      url: url
    )
  end
```

**Comment by zaru** [comment]
@subhashsaran Please tell us the version you are using of firefox and webpush.
If there is a detailed error message, I might be able to advice.

**Comment by sidereel-dev** [comment]
I'm having the same problem, using FireFox 49.0.1 and WebPush 0.2.4,  the values sent by a chrome subscription object work but there is a decoding problem with firefox:

irb(main):032:0> firefox.endpoint
=> "https://updates.push.services.mozilla.com/wpush/v1/gAAAAABYBX8MY9yEa7tXRt0vHjd7R7GgUOKVayEmov3GTMjXdto-E2Z-qIEB3Zwj8fCdSl10TxT2erJNi6WMnrjIqcSuIeGQiLZ3CUhgaRXWZE0LnWEPGWhe1iyOGF4eSUZl3Oe3Qrj3"
irb(main):033:0> firefox.auth
=> "1C1PBkJQsVwD9tkuLR1x5A"
irb(main):034:0> firefox.p256dh
=> "BK74n-ZA6kfMDEuCFbQH1Y5T33p39PvnzNeuD5LqTs8cF-uaQFUHn_v5kwV6dYIIL4nFabxghQNF_vlnAXX7OiU"
irb(main):035:0> Webpush.payload_send({ message: JSON.generate({body: 'firefox push test'}), endpoint: firefox.endpoint, p256dh: firefox.p256dh, auth: firefox.auth, api_key: api_key})                                                                                      
ArgumentError: invalid base64
        from /opt/ruby-2.2.0/lib/ruby/2.2.0/base64.rb:73:in `unpack'
        from /opt/ruby-2.2.0/lib/ruby/2.2.0/base64.rb:73:in `strict_decode64'
        from /opt/ruby-2.2.0/lib/ruby/2.2.0/base64.rb:89:in `urlsafe_decode64'
        from /data/sidereel/shared/bundle/ruby/2.2.0/gems/webpush-0.2.4/lib/webpush/encryption.rb:16:in `encrypt'
        from /data/sidereel/shared/bundle/ruby/2.2.0/gems/webpush-0.2.4/lib/webpush.rb:42:in `build_payload'
        from /data/sidereel/shared/bundle/ruby/2.2.0/gems/webpush-0.2.4/lib/webpush.rb:32:in `payload_send'
        from (irb):35

**Comment by rossta** [comment]
@sidereel-dev Ah, I think I got it now. I was able to reproduce the issue with your data.

Both the `p256dh` and `auth` ("BK74n-ZA6kfMDEuCFbQH1Y5T33p39PvnzNeuD5LqTs8cF-uaQFUHn_v5kwV6dYIIL4nFabxghQNF_vlnAXX7OiU", "1C1PBkJQsVwD9tkuLR1x5A", respectively) cannot be decoded with `Base64.urlsafe_decode64` in Ruby:

```ruby
irb(main):> Base64.urlsafe_decode64(p256dh)
ArgumentError: invalid base64
irb(main):> Base64.urlsafe_decode64(auth)
ArgumentError: invalid base64
```

In Ruby 2.2, the string length must be divisible by 4. To ensure this, we often will see the encoding step add padding to the end of the encoded string as `"="` characters. 

Both of the keys in your FireFox subscription are missing padding and neither are divisible by 4:

```ruby
irb(main)> p256dh.length
=> 87
irb(main)> auth.length
=> 22
```

Supplying the proper padding for both with allow the decoding to succeed:

```ruby
irb(main)> Base64.urlsafe_decode64(p256dh + "=")
=> "\x04\xAE\xF8\x9F\xE6@\xEAG\xCC\fK\x82\x15\xB4\a\xD5\x8ES\xDFzw\xF4\xFB\xE7\xCC\xD7\xAE\x0F\x92\xEAN\xCF\x1C\x17\xEB\x9A@U\a\x9F\xFB\xF9\x93\x05zu\x82\b/\x89\xC5i\xBC`\x85\x03E\xFE\xF9g\x01u\xFB:%"
irb(main)> Base64.urlsafe_decode64(auth + "==")
=> "\xD4-O\x06BP\xB1\\\x03\xF6\xD9.-\x1Dq\xE4"
```

Note the [Ruby 2.3 docs for `Base64`](https://ruby-doc.org/stdlib-2.3.1/libdoc/base64/rdoc/Base64.html) provide the padding before performing the strict decoding:

```ruby
# File base64.rb, line 97
def urlsafe_decode64(str)
  # NOTE: RFC 4648 does say nothing about unpadded input, but says that
  # "the excess pad characters MAY also be ignored", so it is inferred that
  # unpadded input is also acceptable.
  str = str.tr("-_", "+/")
  if !str.end_with?("=") && str.length % 4 != 0
    str = str.ljust((str.length + 3) & ~3, "=")
  end
  strict_decode64(str)
end
```

That means in Ruby 2.3, your Firefox subscription will work!

So, that leaves you with the option of upgrading to Ruby 2.3 and/or we could also patch `webpush` with the Ruby 2.3 code to ensure `p256dh` and `auth` keys are correctly decoded in Ruby 2.2.

**Comment by subhashsaran** [comment]
Hey,

How i solved it was I replaced Ruby 2.1 base64.rb file with Ruby 2.3 base64.rb and everything worked fine.

**Comment by sidereel-dev** [comment]
Thanks for the response!   We'll look into bumping our ruby to 2.3.
